modelforge/dataset/dataset.py

```diff
@@ -459,16 +672,48 @@ def _load_or_process_data(
             The HDF5 dataset instance to use.
         """
 
-        # if not cached, download and process
-        if not os.path.exists(data.processed_data_file):
-            if not os.path.exists(data.raw_data_file):
-                data._download()
-            # load from hdf5 and process
+        # For efficiency purposes, we first want to see if there is an npz file available before reprocessing the hdf5
+        # file, expanding the gzziped archive or download the file.
+        # Saving to cache will create an npz file and metadata file.
+        # The metadata file will contain the keys used to generate the npz file, the checksum of the hdf5 and gz
+        # file used to generate the npz file.  We will look at the metadata file and compare this to the
+        # variables saved in the HDF5Dataset class to determine if the npz file is valid.
+        # It is important to check the keys used to generate the npz file, as these are allowed to be changed by the user.
+
+        if data._file_validation(
+            data.processed_data_file["name"],
+            data.local_cache_dir,
+        ) and (
+            data._metadata_validation(
+                data.processed_data_file["name"].replace(".npz", ".json"),
+                data.local_cache_dir,
+            )
+            and not data.force_download
+            and not data.regenerate_cache
+        ):
+
+            data._from_file_cache()
```

Comment:
very minor comment: `data._from_file_cache() could be moved outside the control flow since each of the branches is calling this method

Reply:
It could be, but I like having it in each as I think it makes it clearer the exact flow being executed for each conditional block (i.e., keeps the logic self-contained and localized in the file).